File: 08_tzwitter_app/app/src/lib/tweet.ts
export interface CollectedData {
  level: number;
  mintableDate: Date;
}

export interface Tweet {
  id: number;
  author: string;
  content: Content;
  likes: number;
  isLiked: boolean;
  collected?: CollectedData;
}

export type Content = Image | Text;

export interface Image {
  type: "image",
  data: string
}
export interface Text {
  type: "text",
  data: string
}

File: 08_tzwitter_app/kernel/src/core/tweet.rs
// SPDX-FileCopyrightText: 2023 Marigold <contact@marigold.dev>
//
// SPDX-License-Identifier: MIT

use serde::Serialize;

use crate::core::public_key_hash::PublicKeyHash;

use super::message::PostTweet;

#[derive(Serialize)]
pub struct Tweet {
    pub author: PublicKeyHash,
    pub content: TweetContent,
    pub likes: u64,
}

#[derive(Serialize)]
pub enum TweetContent {
    Image(Vec<u8>),
    Text(String),
}

impl From<PostTweet> for Tweet {
    fn from(post_tweet: PostTweet) -> Self {
        let PostTweet { author, content } = post_tweet;
        let tweet_content = TweetContent::Text(content);
        Tweet {
            author,
            content: tweet_content,
            likes: 0,
        }
    }
}

impl Tweet {
    pub fn like(self) -> Self {
        Self {
            likes: self.likes + 1,
            ..self
        }
    }
}
